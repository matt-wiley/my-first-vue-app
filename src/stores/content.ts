import { defineStore } from "pinia";

import type Source from "@/models/source";
import type ArticleRecord from "@/models/articleRecord";
import type Article from "@/models/article";
import type { SourceRecord } from "@/models/sourceRecord";
import HashUtils, { HashAlgo } from "@/utils/hashUtils";
import StringUtils from "@/utils/stringUtils";

export const useContentStore = defineStore({
  id: "content",

  state: () => ({
    sources: [] as SourceRecord[],
    articles: [] as ArticleRecord[],
  }),

  getters: {
    getAllSources: (state) => state.sources,
    getAllArticles: (state) => state.articles,
    getArticles: (state) => {
      return state.articles.filter((a) => !a.isTombstoned);
    },
    getArticlesForSourceId: (state) => (sourceId: string) => {
      return state.articles
        .filter((a) => !a.isTombstoned)
        .filter((a) => a.sourceId === sourceId);
    },
    getArticle: (state) => (articleSha: string | undefined) => {
      if (!articleSha) return undefined;
      return state.articles.find((a) => a.id === articleSha);
    },
  },

  actions: {
    async addSource(source: Source): Promise<SourceRecord> {
      const sourceRecord = { ...source } as SourceRecord;
      if (StringUtils.isEmpty(sourceRecord.feedUrl)) {
        throw new Error("Feed URL is required");
      }
      // @ts-ignore Typescript doesn't know about the guard clause above
      const hash = await HashUtils.digest(HashAlgo.SHA1, sourceRecord.feedUrl);
      sourceRecord.id = `S-${hash}`;
      this.sources.push(sourceRecord);
      return sourceRecord;
    },
    async addArticle(sourceRecord: SourceRecord, article: Article): Promise<ArticleRecord> {
      const articleRecord = { ...article } as ArticleRecord;
      articleRecord.sourceId = sourceRecord.id;

      const sha = await HashUtils.digest(
        HashAlgo.SHA256,
        `${articleRecord.sourceId}${articleRecord.title}${articleRecord.link}`
      );
      const id = await HashUtils.digest(
        HashAlgo.SHA1,
        sha
      );
      articleRecord.sha = sha;
      articleRecord.id = `A-${id}`;
      this.articles.push(articleRecord);
      return articleRecord;
    },
    async deleteArticle(articleRecord: ArticleRecord) {
      if (articleRecord.freshness !== "Stale") {
        articleRecord.isTombstoned = true;
      }
      else {
        const index = this.articles.indexOf(articleRecord);
        if (index > -1) {
          this.articles.splice(index, 1);
        }
      }
    },
  },
});
