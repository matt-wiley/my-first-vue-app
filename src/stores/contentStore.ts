import InMemoryContentStore from "./inMemoryContentStore";


export const contentStore = InMemoryContentStore.getInstance();