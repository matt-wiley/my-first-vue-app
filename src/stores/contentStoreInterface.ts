import type ArticleEntity from "@/models/articleEntity";
import type SourceEntity from "@/models/sourceEntity";
import type { Maybe } from "@/types/maybe";
import type { StoreDefinition } from "pinia";

export default interface ContentStoreInterface {

  // Store API
  getReactiveContentStore(): any;
  getStoreId(): string;
  clearStore(): void;

  // Sources API
  getAllSources(): SourceEntity[];
  getSourcesCount(): Promise<number>; // returns the number of sources in the store
  getSourceById(id: string): Promise<Maybe<SourceEntity>>;
  addSource(source: Partial<SourceEntity>): Promise<SourceEntity>;
  updateSource(source: Partial<SourceEntity>): Promise<SourceEntity>;
  deleteSource(id: string): Promise<void>; 
  deleteAllSources(): Promise<number>; // returns the number of sources deleted

  // Articles API
  getAllArticles(): ArticleEntity[];
  getAllArticlesCount(): Promise<number>; // returns the number of articles in the store
  getArticleById(id: string): Promise<Maybe<ArticleEntity>>;
  addArticle(article: Partial<ArticleEntity>): Promise<ArticleEntity>;
  updateArticle(article: ArticleEntity): Promise<ArticleEntity>;
  deleteArticle(id: string): Promise<void>;
  deleteAllArticles(): Promise<number>; // returns the number of articles deleted
  getArticlesForSourceId(sourceId: string): Promise<ArticleEntity[]>;
  getArticlesCountForSourceId(sourceId: string): Promise<number>; // returns the number of articles for the source for the given sourceId

  // Refresh API
  refreshSource(sourceId: string): Promise<number>; // returns the number of articles added
  refreshAllSources(): Promise<number>; // returns the number of articles added

}