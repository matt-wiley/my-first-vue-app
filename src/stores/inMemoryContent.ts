import type Article from "@/models/article";
import type ArticleRecord from "@/models/articleRecord";
import type Source from "@/models/source";
import type { SourceRecord } from "@/models/sourceRecord";
import HashUtils, { HashAlgo } from "@/utils/hashUtils";
import StringUtils from "@/utils/stringUtils";
import { defineStore } from "pinia";

export const useInMemoryContentStore = defineStore({
  id: "content",

  state: () => ({
    sources: [],
    articles: [],
  }),

  getters: {
    getAllSources: (state) => state.sources,
    getSource: (state) => (sourceId: string | undefined) => {
      if (!sourceId) return undefined;
      return state.sources.find((s) => s.id === sourceId);
    },
    getAllArticles: (state) => state.articles,
    getArticles: (state) => {
      return state.articles.filter((a) => !a.isTombstoned);
    },
    getArticlesForSourceId: (state) => (sourceId: string) => {
      return state.articles
        .filter((a) => !a.isTombstoned)
        .filter((a) => a.sourceId === sourceId);
    },
    getArticle: (state) => (articleSha: string | undefined) => {
      if (!articleSha) return undefined;
      return state.articles.find((a) => a.id === articleSha);
    },
  },

  actions: {
    async addSource(source: Source): Promise<SourceRecord> {
      const sourceRecord = { ...source } as SourceRecord;
      if (StringUtils.isEmpty(sourceRecord.feedUrl)) {
        throw new Error("Feed URL is required");
      }
      // @ts-ignore Typescript doesn't know about the guard clause above
      const hash = await HashUtils.digest(HashAlgo.SHA1, sourceRecord.feedUrl);
      sourceRecord.id = `S-${hash}`;
      this.sources.push(sourceRecord);
      return sourceRecord;
    },
    async deleteSource(sourceRecord: SourceRecord) {
      const index = this.sources.indexOf(sourceRecord);
      if (index > -1) {
        this.sources.splice(index, 1);
      }
      const articles = this.articles.filter((a) => a.sourceId === sourceRecord.id);
      for (const article of articles) {
        this.deleteArticle(article);
      }
    },
    async deleteSourceById(sourceId: string) {
      const sourceRecord = this.sources.find((s) => s.id === sourceId);
      if (sourceRecord) {
        this.deleteSource(sourceRecord);
      }
    },
    async addArticle(sourceRecord: SourceRecord, article: Article): Promise<ArticleRecord> {
      const articleRecord = { ...article } as ArticleRecord;
      articleRecord.sourceId = sourceRecord.id;

      const sha = await HashUtils.digest(
        HashAlgo.SHA256,
        `${articleRecord.sourceId}${articleRecord.title}${articleRecord.link}`
      );
      const id = await HashUtils.digest(
        HashAlgo.SHA1,
        sha
      );
      articleRecord.sha = sha;
      articleRecord.id = `A-${id}`;
      this.articles.push(articleRecord);
      return articleRecord;
    },
    async deleteArticle(articleRecord: ArticleRecord) {
      if (articleRecord.freshness !== "Stale") {
        articleRecord.isTombstoned = true;
      }
      else {
        const index = this.articles.indexOf(articleRecord);
        if (index > -1) {
          this.articles.splice(index, 1);
        }
      }
    },
    async deleteArticleById(articleId: string) {
      const articleRecord = this.articles.find((a) => a.id === articleId);
      if (articleRecord) {
        this.deleteArticle(articleRecord);
      }
    },
  },
});
