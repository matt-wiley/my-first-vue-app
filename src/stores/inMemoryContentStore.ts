import type ArticleEntity from "@/models/articleEntity";
import type SourceEntity from "@/models/sourceEntity";
import type ContentStoreInterface from "@/stores/contentStoreInterface";
import type { Maybe } from "@/types/maybe";
import HashUtils, { HashAlgo } from "@/utils/hashUtils";
import ValidationUtils from "@/utils/validationUtils";
import { get } from "http";
import { defineStore } from "pinia";



class ContentStoreUtils {

  static async generateSourceId(source: Partial<SourceEntity>): Promise<string> {
    if (ValidationUtils.isEmptyString(source.feedUrl)) {
      throw new Error("source.feedUrl cannot be null or undefined");
    }
    return `S-${await HashUtils.digest(HashAlgo.SHA1, source.feedUrl! )}`
  }


  static async generateArticleShaAndId(article: Partial<ArticleEntity>): Promise<{ sha: string, id: string }> {
    if (ValidationUtils.isEmptyString(article.sourceId)) {
      throw new Error("article.sourceId cannot be null or undefined");
    }
    const secondaryInput = article.externalId ? article.externalId : article.title;
    if (ValidationUtils.isEmptyString(secondaryInput)) {
      throw new Error("article.externalId or article.title must be set");
    }
    const sha = await HashUtils.digest(HashAlgo.SHA256, article.sourceId + "" + secondaryInput);
    const id = `A-${await HashUtils.digest(HashAlgo.SHA1, sha)}`;
    return { sha, id };
  }

}


const useInMemoryContentStore = defineStore({
  id: "inMemoryContentStore",
  state: () => ({
    sources: [],
    articles: [],
  }),
  getters: {
    getAllSources: (state) => state.sources,
    getSourceById: (state) => (id: string) => state.sources.find((source: SourceEntity) => source.id === id),
    getSourcesCount: (state) => state.sources.length,
    findSourceWithFeedUrl: (state) => (feedUrl: string) => state.sources.find((source: SourceEntity) => source.feedUrl === feedUrl),
    getAllArticles: (state) => state.articles,
    getArticleById: (state) => (id: string) => state.articles.find((article: ArticleEntity) => article.id === id),
  },
  actions: {
    /// ===========================================================================================
    /// 
    ///   Mutators for SourceEntity
    ///
    async addSource(source: Partial<SourceEntity>) {
      if (ValidationUtils.isEmptyString(source.title)) {
        throw new Error("source.title cannot be null or undefined");
      }
      if (ValidationUtils.isEmptyString(source.feedUrl)) {
        throw new Error("source.feedUrl cannot be null or undefined");
      }
      if (this.findSourceWithFeedUrl(source.feedUrl!)) {
        throw new Error("source.feedUrl must be unique");
      }

      source.id = await ContentStoreUtils.generateSourceId(source);

      // @ts-ignore: This expression is not callable. Type 'never[]' has no call signatures. ts(2349)
      this.sources.push(source);
      return source as SourceEntity
    },
    updateSource(source: Partial<SourceEntity>) {
      if (ValidationUtils.isEmptyString(source.id)) {
        throw new Error("source.id cannot be null or undefined");
      }
      const existingSource = this.getSourceById(source.id!);
      if(!existingSource) {
        throw new Error(`Source with id ${source.id} not found`);
      }
      Object.assign(existingSource, source);
      return existingSource;
    },
    deleteSource(id: string) {
      const existingSource = this.getSourceById(id)
      if(!existingSource) {
        throw new Error(`Source with id ${id} not found`)
      }
      this.sources = this.sources.filter((source: SourceEntity) => source.id !== id);
    },
    deleteAllSources() {
      this.sources = [];
      return this.sources.length;
    },

    /// ===========================================================================================
    ///
    ///   Mutators for ArticleEntity
    ///
    async addArticle(article: Partial<ArticleEntity>) {

      if (ValidationUtils.isEmptyString(article.sourceId)) {
        throw new Error("article.sourceId cannot be null or undefined");
      }
      const secondaryInput = article.externalId ? article.externalId : article.title;
      if (ValidationUtils.isEmptyString(secondaryInput)) {
        throw new Error("article.externalId or article.title must be set");
      }

      const { sha, id } = await ContentStoreUtils.generateArticleShaAndId(article);

      article.id = id;
      article.sha = sha;

      // @ts-ignore: This expression is not callable. Type 'never[]' has no call signatures. ts(2349)
      this.articles.push(article);
      return article as ArticleEntity;
    },
    updateArticle(article: Partial<ArticleEntity>) {
      // TODO: updateArticle may need some more restrictions around what can be updated
      if (ValidationUtils.isEmptyString(article.id)) {
        throw new Error("article.id cannot be null or undefined");
      }
      const existingArticle = this.getArticleById(article.id!);
      if(!existingArticle) {
        throw new Error(`Article with id ${article.id} not found`);
      }
      Object.assign(existingArticle, article);
      return existingArticle;
    },


    /// ===========================================================================================
    ///
    ///   Global Mutators
    ///
    clearStore() {
      this.deleteAllSources()
      this.articles = [];
    }
  },});


export default class InMemoryContentStore implements ContentStoreInterface {

  private static INSTANCE: InMemoryContentStore;
  private readonly piniaContentStore;

  private constructor() {
    this.piniaContentStore = useInMemoryContentStore;
  }
  
  public static getInstance(): InMemoryContentStore {
    if (!InMemoryContentStore.INSTANCE) {
      InMemoryContentStore.INSTANCE = new InMemoryContentStore();
    }
    return InMemoryContentStore.INSTANCE;
  }

  // 
  // Store API
  //

  getReactiveContentStore(): any {
    return this.piniaContentStore();
  }
  getStoreId(): string {
    return this.piniaContentStore().$id;
  }
  clearStore(): void {
    this.getReactiveContentStore().clearStore();
  }

  //
  // Sources API
  //

  addSource(source: Partial<SourceEntity>): Promise<SourceEntity> {
    return Promise.resolve(this.getReactiveContentStore().addSource(source));
  }
  getSourceById(id: string): Promise<Maybe<SourceEntity>> {
    return Promise.resolve(this.getReactiveContentStore().getSourceById(id));
  }
  updateSource(source: Partial<SourceEntity>): Promise<SourceEntity> {
    return Promise.resolve(this.getReactiveContentStore().updateSource(source));
  }
  deleteSource(id: string): Promise<void> {
    return Promise.resolve(this.getReactiveContentStore().deleteSource(id));
  }
  getAllSources(): SourceEntity[] {
    // @ts-ignore: This expression is not callable. Type 'never[]' has no call signatures. ts(2349)
    return this.getReactiveContentStore().getAllSources
  }
  getSourcesCount(): Promise<number> {
    return Promise.resolve(this.getReactiveContentStore().getSourcesCount);
  }
  deleteAllSources(): Promise<number> {
    return Promise.resolve(this.getReactiveContentStore().deleteAllSources());
  }

  //
  // Articles API
  //

  addArticle(article: Partial<ArticleEntity>): Promise<ArticleEntity> {
    return Promise.resolve(this.getReactiveContentStore().addArticle(article));
  }
  getArticleById(id: string): Promise<Maybe<ArticleEntity>> {
    return Promise.resolve(this.getReactiveContentStore().getArticleById(id));
  }
  updateArticle(article: Partial<ArticleEntity>): Promise<ArticleEntity> {
    return Promise.resolve(this.getReactiveContentStore().updateArticle(article));
  }
  deleteArticle(id: string): Promise<void> {
    // TODO: // YAH // Implement deleteArticle 
    throw new Error("Method not implemented.");
  }
  getAllArticles(): ArticleEntity[] {
    return this.getReactiveContentStore().getAllArticles;
  }
  getAllArticlesCount(): Promise<number> {
    throw new Error("Method not implemented.");
  }
  deleteAllArticles(): Promise<number> {
    throw new Error("Method not implemented.");
  }
  getArticlesForSourceId(sourceId: string): Promise<ArticleEntity[]> {
    throw new Error("Method not implemented.");
  }
  getArticlesCountForSourceId(sourceId: string): Promise<number> {
    throw new Error("Method not implemented.");
  }

  //
  // Refresh API
  //

  refreshSource(sourceId: string): Promise<number> {
    throw new Error("Method not implemented.");
  }
  refreshAllSources(): Promise<number> {
    throw new Error("Method not implemented.");
  }
}
