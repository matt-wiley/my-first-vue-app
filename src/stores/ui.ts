import { defineStore } from "pinia";
import { useLocalStorage } from "@vueuse/core"

type OptionalString = string | undefined

const _uiStateStoreDefinition = (() => {

    const KEY_SELECTED_SOURCE_ID: string = "mfva-rss.ui.selectedSourceId"
    const KEY_SELECTED_ARTICLE_ID: string  = "mfva-rss.ui.selectedArticleId"

    return defineStore({
        id: "ui",

        hydrate: (currentState, initialState) => {
            // @ts-expect-error: https://github.com/microsoft/TypeScript/issues/43826
            currentState.selectedSourceId = useLocalStorage<OptionalString>(KEY_SELECTED_SOURCE_ID, undefined)
            // @ts-expect-error: https://github.com/microsoft/TypeScript/issues/43826
            currentState.selectedArticleId = useLocalStorage<OptionalString>(KEY_SELECTED_ARTICLE_ID, undefined)
        },

        state: () => ({
            selectedSourceId: useLocalStorage<OptionalString>(KEY_SELECTED_SOURCE_ID, undefined),
            selectedArticleId: useLocalStorage<OptionalString>(KEY_SELECTED_ARTICLE_ID, undefined),
        }),
    
        getters: {
            getSelectedSourceId: (state) => state.selectedSourceId,
            getSelectedArticleId: (state) => state.selectedArticleId,
        },
    
        actions: {
            setSelectedSourceId(sourceId?: string) {
                this.selectedSourceId = sourceId;
            },
            setSelectedArticleId(articleId?: string) {
                this.selectedArticleId = articleId;
            },
        },
    });

})();


export const useUIStateStore = _uiStateStoreDefinition
