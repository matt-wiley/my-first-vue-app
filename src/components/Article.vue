<script setup lang="ts">
import type ArticleRecord from "@/models/articleRecord";
import { useContentStore } from "@/stores/content";
import { useUIStateStore } from "@/stores/ui";
import DateUtils from "@/utils/dateUtils";

const uiState = useUIStateStore();
const content = useContentStore();

function getSelectedArticle() {
  return content.getArticle(uiState.getSelectedArticleId) as ArticleRecord | null;
}

function presentDate(date: Date | null | undefined) {
  if (!date) {
    return "";
  }
  return DateUtils.presentDateAsMMMM_DD_YYYY_HH_MM_AMPM(date);
}

</script>

<template>
  <div class="ma0 pa0 w-100 h-100 overflow-x-hidden overflow-y-scroll" v-bind:articleId="uiState.selectedArticleId">
    <section class="" v-if="uiState.selectedArticleId">
      <div class="br2 ml4 mt5 mr4 pa4 bg-white h-100">
        <h2 id="article-title" class="pb2">{{ getSelectedArticle()?.title }}</h2>
        <div class="pb4">
          <p id="article-author" v-html="getSelectedArticle()?.author" />
          <p id="article-date">
            {{ presentDate(getSelectedArticle()?.date) }}
          </p>
        </div>
        <div id="article-content" v-html="getSelectedArticle()?.content" />
        <div class="mt4">
          <!-- TODO: DO NOT TEST -- Remove this, used for development only-->
          <p>sha = {{ getSelectedArticle()?.sha }}</p>
          <p>id = {{ getSelectedArticle()?.id }}</p>
          <p>sourceId = {{ getSelectedArticle()?.sourceId }}</p>
          <p>isTombstoned = {{ getSelectedArticle()?.isTombstoned }}</p>
          <p>freshness = {{ getSelectedArticle()?.freshness }}</p>
        </div>
        <div id="article-footer-whitespace" class="vh-25 w-100">&nbsp;</div>
      </div>
    </section>
    <section v-else>
      <div class="br2 ml4 mt5 mr4 pa4 h-100">
        <p id="no-article" class="tc">No article selected</p>
      </div>
    </section>
  </div>
</template>

<style scoped></style>
