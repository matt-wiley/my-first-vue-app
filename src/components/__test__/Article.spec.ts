
import { mount } from "@vue/test-utils";
import { describe, expect, it, beforeEach } from "vitest";
import { useContentStore } from "@/stores/content";
import { useUIStateStore } from "@/stores/ui";
import SampleDataUtils from "@/utils/sampleDataUtils";
import StringUtils from "@/utils/stringUtils";
import Article from "../Article.vue";
import { createPinia, setActivePinia } from "pinia";
import DateUtils from "@/utils/dateUtils";


const getFeedUrlForTest = () => {
  return "https://www.example.com/" + StringUtils.randomStringOfLength(20);
}

const setupContentStore = async (piniaForTest: any) => {
  // This sets up the following scenario:
  // 
  // contentStore:
  //   sourceARecord:
  //     articleARecord: active
  //     articleBRecord: tombstoned
  //   sourceBRecord:
  //     articleCRecord: active

  const contentStore = useContentStore(piniaForTest);
  contentStore.sources = [];
  contentStore.articles = [];

  const feedAUrlForTest = getFeedUrlForTest();
  const feedBUrlForTest = getFeedUrlForTest();

  const sourceA = SampleDataUtils.generateSource();
  sourceA.feedUrl = feedAUrlForTest;
  const sourceARecord = await contentStore.addSource(sourceA);

  const sourceB = SampleDataUtils.generateSource();
  sourceB.feedUrl = feedBUrlForTest;
  const sourceBRecord = await contentStore.addSource(sourceB);

  const articleA = SampleDataUtils.generateArticle();
  const articleARecord = await contentStore.addArticle(sourceARecord, articleA);

  const articleB = SampleDataUtils.generateArticle();
  const articleBRecord = await contentStore.addArticle(sourceARecord, articleB);
  contentStore.deleteArticle(articleBRecord);

  const articleC = SampleDataUtils.generateArticle();
  const articleCRecord = await contentStore.addArticle(sourceBRecord, articleC);

  return {
    contentStore: contentStore,
    sourceARecord: sourceARecord,
    sourceBRecord: sourceBRecord,
    articleARecord: articleARecord,
    articleBRecord: articleBRecord,
    articleCRecord: articleCRecord
  };
}

const setupUiStore = async (piniaForTest: any, selectedArticleId: string | undefined = undefined) => {
  const uiStateStore = useUIStateStore(piniaForTest);
  uiStateStore.setSelectedArticleId(selectedArticleId);
  return {
    uiStateStore: uiStateStore
  }
}



describe("Article.vue", () => {
  
  it("renders properly when an article is selected", async () => {
    
    const piniaForTest = setActivePinia(createPinia());
    const contentStoreSetup = await setupContentStore(piniaForTest);
    await setupUiStore(piniaForTest, contentStoreSetup.articleARecord.id);

    const wrapper = mount(Article, {});
    
    // Article section should be rendered
    expect(wrapper.find("section").exists()).toBe(true);
    
    // Article content should be rendered
    expect(wrapper.find("#article-title").exists()).toBe(true);
    expect(wrapper.find("#article-title").element.tagName).toEqual("H2");
    expect(wrapper.find("#article-title").text()).toBe(contentStoreSetup.articleARecord.title);

    expect(wrapper.find("#article-author").exists()).toBe(true);
    expect(wrapper.find("#article-author").element.tagName).toEqual("P");
    expect(wrapper.find("#article-author").element.innerHTML).toBe(contentStoreSetup.articleARecord.author);

    expect(wrapper.find("#article-date").exists()).toBe(true);
    expect(wrapper.find("#article-date").element.tagName).toEqual("P");

    expect(contentStoreSetup.articleARecord.date).toBeInstanceOf(Date); 
    // @ts-ignore
    const expectedDate = DateUtils.presentDateAsMMMM_DD_YYYY_HH_MM_AMPM(contentStoreSetup.articleARecord.date);
    expect(wrapper.find("#article-date").text()).toBe(expectedDate);

    expect(wrapper.find("#article-content").exists()).toBe(true);
    expect(wrapper.find("#article-content").element.tagName).toEqual("DIV");
    expect(wrapper.find("#article-content").element.innerHTML).toBe(contentStoreSetup.articleARecord.content);

    expect(wrapper.find("#article-footer-whitespace").exists()).toBe(true);
    expect(wrapper.find("#article-footer-whitespace").element.tagName).toEqual("DIV");
    expect(wrapper.find("#article-footer-whitespace").element.innerHTML).toEqual("&nbsp;");

    // No article selected message should not be rendered
    expect(wrapper.find("#no-article").exists()).toBe(false);

  });


  it("renders properly when no article is selected", async () => {

    const piniaForTest = setActivePinia(createPinia());
    await setupUiStore(piniaForTest, undefined);

    expect(useUIStateStore().selectedArticleId).toBe(undefined);

    const wrapper = mount(Article, {});

    // Article section should be rendered
    expect(wrapper.find("section").exists()).toBe(true);
    
    // Article content should not be rendered
    expect(wrapper.find("#article-title").exists()).toBe(false);
    expect(wrapper.find("#article-author").exists()).toBe(false);
    expect(wrapper.find("#article-date").exists()).toBe(false);
    expect(wrapper.find("#article-content").exists()).toBe(false);
    expect(wrapper.find("#article-footer-whitespace").exists()).toBe(false);

    // No article selected message should be rendered
    expect(wrapper.find("#no-article").exists()).toBe(true);
    expect(wrapper.find("#no-article").element.tagName).toEqual("P");
    expect(wrapper.find("#no-article").text()).toBe("No article selected");

  });

});
