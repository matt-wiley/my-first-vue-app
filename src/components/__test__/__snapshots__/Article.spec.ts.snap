// Vitest Snapshot v1

exports[`Article > renders properly 1`] = `
"<div data-v-41fb85bc=\\"\\" articleid=\\"A-3598264bfc61682d74c4d5fd087a75c821778533\\" class=\\"ma0 pa0 w-100 h-100 overflow-x-hidden overflow-y-scroll\\" title=\\"Hello World\\" content=\\"This is a test content\\">
  <section data-v-41fb85bc=\\"\\">
    <div data-v-41fb85bc=\\"\\" class=\\"br2 ml4 mt5 mr4 pa4 bg-white h-100\\">
      <h2 data-v-41fb85bc=\\"\\" class=\\"pb2\\">Gravida arcu mattis diam dignissim ipsum tellus nulla egestas id turpis</h2>
      <div data-v-41fb85bc=\\"\\" class=\\"pb4\\">
        <p data-v-41fb85bc=\\"\\">l31OdU 2RJkg2</p>
        <p data-v-41fb85bc=\\"\\"> May 11, 2021 at 5:09 AM </p>
      </div>
      <div data-v-41fb85bc=\\"\\">Lorem ipsum dolor sit amet consectetur adipiscing elit sed do eiusmod tempor incididunt ut labore et dolore magna aliqua Tellus id interdum velit laoreet id</div>
      <div data-v-41fb85bc=\\"\\" class=\\"mt4\\">
        <p data-v-41fb85bc=\\"\\">sha = d0f5573575489d1eeefe7d7e5aba2ce84eac50be01d5d5761b308efd4b1c8aea</p>
        <p data-v-41fb85bc=\\"\\">id = A-3598264bfc61682d74c4d5fd087a75c821778533</p>
        <p data-v-41fb85bc=\\"\\">sourceId = S-59f983fee7e25adfada792d0cb5cf842b15b2889</p>
        <p data-v-41fb85bc=\\"\\">isTombstoned = </p>
        <p data-v-41fb85bc=\\"\\">freshness = </p>
      </div>
      <div data-v-41fb85bc=\\"\\" class=\\"vh-25 w-100\\">&nbsp;</div>
    </div>
  </section>
</div>"
`;
